File: packages/learning-core/src/learnflow_core/checkpoint_presets.py
"""Scoped entry design over existing checkpoints; never a grader or file writer."""
from typing import Literal
from pathlib import PurePosixPath
from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

PRESET_VERSION = "learnflow.checkpoint-entry.v1"


class RequiredFile(BaseModel):
    model_config = ConfigDict(extra="forbid")
    path: str = Field(min_length=1, max_length=240)
    purpose: str = Field(min_length=2, max_length=600)

    @field_validator("path")
    @classmethod
    def relative_file(cls, value):
        parts = value.split("/")
        if (value != value.strip() or "\\" in value or ":" in value or any(ord(c) < 32 for c in value)
                or PurePosixPath(value).is_absolute() or any(p in {"", ".", "..", ".learnflow", ".git"} for p in parts)
                or any(p == ".env" or p.startswith(".env.") for p in parts)):
            raise ValueError("文件必须是普通项目文件的相对路径，不能包含保护目录或凭据文件")
        return value


class CheckpointPreset(BaseModel):
    model_config = ConfigDict(extra="forbid")
    kind: Literal["knowledge", "overview", "setup", "implementation", "workflow", "deployment", "review"]
    lecture_focus: str = Field(min_length=2, max_length=1200)
    practice_focus: str = Field(default="", max_length=800)
    workflow_step: str = Field(default="", max_length=300)
    required_files: list[RequiredFile] = Field(default_factory=list, max_length=8)

    @model_validator(mode="after")
    def file_contract(self):
        if self.kind == "implementation" and not self.required_files:
            raise ValueError("实现关卡需要明确待完成文件及各自职责")
        paths = [item.path.casefold() for item in self.required_files]
        if len(paths) != len(set(paths)):
            raise ValueError("同一关卡的文件路径不能重复")
        if self.kind in {"knowledge", "overview"} and not self.practice_focus.strip():
            raise ValueError("知识与综述关卡需要明确习题或小练习目标")
        if self.kind == "workflow" and not self.workflow_step.strip():
            raise ValueError("实践关卡需要明确对应的工作流程步骤")
        return self


class PlannedCheckpoint(BaseModel):
    model_config = ConfigDict(extra="forbid")
    key: str = Field(min_length=1, max_length=80, pattern=r"^[a-zA-Z0-9_-]+$")
    title: str = Field(min_length=2, max_length=255)
    objective: str = Field(min_length=2, max_length=1200)
    entry_preset: CheckpointPreset


def validate_mode_preset(mode, preset, order):
    if preset is None:  # Existing clients and confirmed routes remain readable.
        return
    if mode == "learning" and preset.kind != "knowledge":
        raise ValueError("学习型项目按知识递进规划，每关配讲义与一份习题")
    if mode == "practice" and order == 1 and preset.kind != "overview":
        raise ValueError("实践型项目第一关必须是综述，包含总纲讲义与小练习")
    if mode == "practice" and order > 1 and preset.kind not in {"workflow", "review"}:
        raise ValueError("实践型后续关卡应对应一个工作流程步骤")
    if mode == "experiment" and preset.kind not in {"setup", "implementation", "deployment", "review"}:
        raise ValueError("实验型项目以实现关卡组成交付，可包含环境、部署和复核关卡")


def checkpoint_entry_preset(project, checkpoint, stage=None, *, workflow_titles=()):
    stage = stage or {}
    mode = project.project_mode or "learning"
    raw = (checkpoint.brief or {}).get("entry_preset") or stage.get("entry_preset")
    title = checkpoint.title
    objective = checkpoint.description or title
    if raw:
        preset = CheckpointPreset.model_validate(raw).model_dump()
    else:
        key = str((checkpoint.brief or {}).get("checkpoint_key") or stage.get("key") or "")
        kind = ("knowledge" if mode == "learning" else "overview" if mode == "practice" and checkpoint.order == 1
                else "workflow" if mode == "practice" else "setup" if key in {"define", "setup", "environment"}
                else "review" if key in {"reflect", "review"} else "deployment" if key == "deploy" else "implementation")
        files = [{"path": f["path"], "purpose": f.get("reason") or objective} for f in stage.get("related_files", [])
                 if f.get("role") in {"implementation", "test", "docs"}]
        preset = {"kind": kind, "lecture_focus": objective,
                  "practice_focus": ("围绕本关知识完成一份习题，先理解再应用。" if mode == "learning" else
                                     "用少量小练习检查工作目标、整体流程、输入输出与关键风险。" if kind == "overview" else ""),
                  "workflow_step": title if mode == "practice" else "", "required_files": files}
    overview = preset["kind"] == "overview"
    return {"schema_version": PRESET_VERSION, **preset, "configured": bool(raw),
            "overview_outline": " → ".join(workflow_titles)[:2000] if overview else "",
            "lecture_title": f"{project.name} · 工作总纲" if overview else f"{title} · 讲义",
            "practice_title": "综述小练习" if overview else f"{title} · 习题",
            "file_kinds": ["lecture", "practice"] if preset["practice_focus"] else ["lecture"],
            "help_surface": "code_paper", "desktop_guidance": mode in {"experiment", "practice"},
            "needs_file_plan": preset["kind"] == "implementation" and not preset["required_files"]}


def generation_focus(preset):
    return (f"本关类型：{preset['kind']}。讲义重点：{preset['lecture_focus']}。\n"
            + (f"习题集目标：{preset['practice_focus']}。\n" if preset["practice_focus"] else "")
            + (f"总纲需覆盖完整工作流程：{preset['overview_outline']}。\n" if preset.get("overview_outline") else "")
            + (f"工作流程步骤：{preset['workflow_step']}。\n" if preset["workflow_step"] else "")
            + "待完成文件（仅作为任务约束，不生成完整答案）："
            + "；".join(f"{f['path']}：{f['purpose']}" for f in preset["required_files"]))

File: packages/learning-client/src/project-guidance/checkpoint-presets.ts
/** Entry design is a content plan. Files exist only after the scoped file API succeeds. */
export type CheckpointPreset = {
  kind: 'knowledge' | 'overview' | 'setup' | 'implementation' | 'workflow' | 'deployment' | 'review'
  lecture_focus: string; practice_focus: string; workflow_step: string
  required_files: Array<{ path: string; purpose: string }>
}
export type CheckpointEntryPreset = CheckpointPreset & {
  schema_version: 'learnflow.checkpoint-entry.v1'; configured: boolean
  overview_outline?: string; lecture_title: string; practice_title: string; file_kinds: Array<'lecture' | 'practice'>
  help_surface: 'code_paper'; desktop_guidance: boolean; needs_file_plan: boolean
}
export const CHECKPOINT_PLANNING_GUIDANCE = '学习型按知识先备循序递进，每关设计讲义与一份习题。实验型围绕最终交付按组件/接口拆出多个实现关卡；环境准备、部署、复核可单列，每个实现关卡明确若干原子文件路径及职责。实践型第一关是综述，配总纲讲义与小练习；后续每关对应一个工作流程步骤。entry_preset 是待确认设计，不能声称文件已创建；四档帮助只在代码纸内选择。'
export const checkpointPresetSchema = {
  type: 'object', additionalProperties: false,
  required: ['kind', 'lecture_focus', 'practice_focus', 'workflow_step', 'required_files'],
  properties: {
    kind: { type: 'string', enum: ['knowledge', 'overview', 'setup', 'implementation', 'workflow', 'deployment', 'review'] },
    lecture_focus: { type: 'string', minLength: 2, maxLength: 1200 },
    practice_focus: { type: 'string', maxLength: 800 }, workflow_step: { type: 'string', maxLength: 300 },
    required_files: { type: 'array', maxItems: 8, items: { type: 'object', additionalProperties: false,
      required: ['path', 'purpose'], properties: { path: { type: 'string', minLength: 1, maxLength: 240, description: '如 src/parser.c；相对路径，不表示文件已经存在' }, purpose: { type: 'string', minLength: 2, maxLength: 600 } } } },
  },
} as const
export function parseCheckpointPreset(value: unknown, allowUnplannedFiles = false): CheckpointPreset {
  const row = value as CheckpointPreset
  const bounded = (value: unknown, min: number, max: number) => typeof value === 'string' && value.trim().length >= min && value.length <= max
  if (!row || typeof row !== 'object' || !checkpointPresetSchema.properties.kind.enum.includes(row.kind)
      || !bounded(row.lecture_focus, 2, 1200) || !bounded(row.practice_focus ?? '', 0, 800)
      || !bounded(row.workflow_step ?? '', 0, 300) || !Array.isArray(row.required_files) || row.required_files.length > 8)
    throw new Error('关卡预设无效：请提供类型、讲义重点、练习目标和待完成文件。')
  const paths = new Set<string>()
  for (const file of row.required_files) {
    if (!bounded(file?.path, 1, 240) || !bounded(file?.purpose, 2, 600) || file.path !== file.path.trim()
      || /[\\:\x00-\x1f]/.test(file.path) || file.path.split('/').some(p => ['', '.', '..', '.learnflow', '.git', '.env'].includes(p) || p.startsWith('.env.'))
      || paths.has(file.path.toLowerCase())) throw new Error('待完成文件需要唯一、安全的相对路径和具体职责。')
    paths.add(file.path.toLowerCase())
  }
  if (row.kind === 'implementation' && !row.required_files.length && !allowUnplannedFiles) throw new Error('实现关卡必须明确待完成的原子文件。')
  if (['knowledge', 'overview'].includes(row.kind) && !row.practice_focus?.trim()) throw new Error('知识与综述关卡必须有习题或小练习目标。')
  if (row.kind === 'workflow' && !row.workflow_step?.trim()) throw new Error('实践关卡必须对应明确的工作流程步骤。')
  return { kind: row.kind, lecture_focus: row.lecture_focus, practice_focus: row.practice_focus || '', workflow_step: row.workflow_step || '',
    required_files: row.required_files.map(file => ({ path: file.path, purpose: file.purpose })) }
}
